### ISSUE TITLE
Add nested sampling to Pints

### ISSUE THREAD

**Issue by ben18785** [pints-team/pints#107]
Rejection sampling, ellipsoidal sampling and, eventually, multiNest.

**Comment by MichaelClerx** [comment]
It needs tests!

**Comment by ben18785** [comment]
I agree. Adding tests for the various functions and samplers is next on my
agenda.

**Comment by MichaelClerx** [comment]
In general though, no new feature should be merged that

1. Doesn't have tests
2. Doesn't show up in the docs at http://pints.readthedocs.io/

**Comment by ben18785** [comment]
How do I get it to show up in the docs?

**Comment by MichaelClerx** [comment]
There's a directory called `docs` that has a bunch of `.rst` ([reStructuredText](http://docutils.sourceforge.net/docs/user/rst/quickstart.html)) files. It's structured into directories like `optimisation` and `mcmc`. The thing to do would be to find the suitable directory, copy a file describing e.g. adaptive covariance mcmc, and copy and modify it. Then update the `index.rst` in the same directory and it should work!

To test it locally, open a console and navigate to the `docs` directory. Then type `make html` and it should build a website without errors. You can then check in your browser to see if the stuff in `docs/html` makes sense.

After you've pushed and everything has been merged into master, readthedocs.io should automatically pick up on it and rebuild our docs. (This last bit is broken a.t.m., see #175)

**Comment by MichaelClerx** [comment]
Sorry we should really set up some kind of docs for contributing to this project!

**Comment by MichaelClerx** [comment]
Probably won't fit in ask/tell, might be useful to be able to do one step at a time?

**Comment by MichaelClerx** [comment]
General outline:

- Sample from the prior
- While:

  - Calculate likelihood for those points
  - Select points with likelihood greater than some threshold

**Comment by MichaelClerx** [comment]
Here's a list of things we still need to do before we can close this ticket:

- [ ] Add unit tests
- [ ] Implement the `Loggable` interface instead of using `print()`
- [ ] Check if the notebooks are as fast as they can be while staying informative #199 
- [ ] Make sure we 'hide' all functions and methods a user doesn't need to see #174
- [ ] Finish the nested sampling example? #130
- [ ] See if there's any way to make it ask-and-tell-ish (probably not?)
- [ ] See if we can optimise / parallelise anything

**Comment by MichaelClerx** [comment]
Have made a start at this in branch: https://github.com/pints-team/pints/tree/i107-nested

**Comment by MichaelClerx** [comment]
What's your thoughts on this @ben18785 ? We still need a notebook for rejection + a fix to the ellipsoidal notebook
